shell/platform/linux/fl_key_event_plugin.cc

```diff
@@ -71,6 +77,40 @@ void fl_key_event_plugin_send_key_event(FlKeyEventPlugin* self,
   int64_t scan_code = event->hardware_keycode;
   int64_t unicodeScalarValues = gdk_keyval_to_unicode(event->keyval);
 
+  // Remove lock states from state mask.
+  guint state = event->state & ~(GDK_LOCK_MASK | GDK_MOD2_MASK);
+
+  // GTK keeps track of the state of the locks themselves, not the "down" state
+  // of the key. Flutter expects the "down" state of the modifier key, so we
+  // keep track of the down state here, and send it to the framework. This code
+  // has the flaw that if a key event is missed due to the app losing focus,
+  // then this state will still think the key is down when it isn't, but that is
+  // no worse than for other keys until we implement the sync/cancel events.
+  //
+  // This is necessary to do here instead of in the framework because Flutter
+  // does modifier key syncing in the framework, and will turn on/off these keys
+  // as being "pressed" whenever the lock is on, which breaks a lot of
+  // interactions.
+  //
+  // TODO(gspencergoog): get rid of this tracked state when we are tracking the
+  // state of all keys and sending sync/cancel events when focus is gained/lost.
+  static bool shift_lock_down = false;
+  static bool caps_lock_down = false;
+  static bool num_lock_down = false;
+  if (event->keyval == GDK_KEY_Num_Lock) {
```

Comment:
Shouldn't the pressing states be used as source-of-truth and make key events sync to them, instead of vice versa? Even if we don't have sync/cancel for now, using down/up events with incorrect timing should be better than desync'd states.

Reply:
I'm not sure I know what you mean. There's little correlation between the lock state and whether the key is pressed at the moment, so I can't use the lock state to decide if a key is down.

The `state` member of the event contains a bit for whether or not the lock is engaged, not whether or not the lock key is pressed.